pkg/pmem-csi-operator/controller/deployment/deployment_controller.go

```diff
@@ -177,10 +177,14 @@ func (r *ReconcileDeployment) Reconcile(request reconcile.Request) (reconcile.Re
 	defer func() {
 		klog.Infof("Updating deployment status....")
 		d.Deployment.Status.LastUpdated = metav1.Now()
-		if statusErr := r.client.Status().Patch(context.TODO(), d.Deployment, patch); statusErr != nil {
+		// Passing copy of CR to patch as the fake client used in tests
+		// will write back the changes to both status and spec.
+		copy := d.Deployment.DeepCopy()
+		if statusErr := r.client.Status().Patch(context.TODO(), copy, patch); statusErr != nil {
```

Comment:
This code is not obvious. Why is `d.Deployment.Status` getting modified?

Reply:
After re-reading this a few times it became clearer.

```
Beware that r.client.Status() as implemented by the fake client during testing will
patch not just the status, but also the spec. We don't want that and therefore
let it work with a copy of the deployment, then later update just the status.
```

Reply:
Is this a fix for a different problem than the other commit?

The commit message describe what the patch does, but not why we want it.

Reply:
`Patch()` is updating the passed `obj` with patch result: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/client.go#L353-L403

Reply:
> Is this a fix for a different problem than the other commit?
> 
The first commit revealed this. The `modified_deployment_under_reconcile` test is failing after the change. This test expects   modified changes are supposed to visible in the next reconcile loop, but operator could not figureout the change because the `status.Pactch()` updated the cached CR, and the next reconcile loop couldn't capture the change.

It was hidden previously because, reconcile used to redeploy both the controller, node driver even if there is no real change in deployment because of comparison always returns all default fields as changes.